auditmiddleware/_api.py

```diff
@@ -520,13 +520,25 @@ def _create_cadf_event(self, project, res_spec, res_id, res_parent_id,
         if not action:
             return None
 
+        # Try to get project_id from request headers
         project_id = request.environ.get('HTTP_X_PROJECT_ID')
-        # If project_id is undefined, look for another variable. This is
-        # added specific to catching delete events from Neutron
+
+        # If no project_id found, try to get it from the request context
         if project_id is None:
+            # Get the adhoc attributes from the request environment
             adhoc_attrs = request.environ.get('webob.adhoc_attrs', {})
-            context = adhoc_attrs.get('context', {})
-            original_resources = context.get('original_resources', [])
+
+            # Get the context - could be dict or RequestContext object
+            context = adhoc_attrs.get('context')
+
+            # Handle both dict and RequestContext object types
+            original_resources = None
+            if isinstance(context, dict):
+                original_resources = context.get('original_resources', [])
+            elif hasattr(context, 'original_resources'):
+                original_resources = context.original_resources
```

Comment:
(optional)

I do prefer duck-typing i.e. we check if we have explicit `original_resources` and everything else that supports `get()` we use `get()` on.
```python
context = adhoc_attrs.get('context', {})

if hasattr(context, 'original_resources'):
    original_resources = context.original_resources
else:
    original_resources = context.get('original_resources', [])
```

We could also do something like
```python
original_resources = getatr(context, 'original_resources', [])
if not original_resources:
    context.get('original_resources', [])
```

Reply:
Makes sense, more pythonic. Appreciate the feedback, and correction.